### PR TITLE
data service - new command line options

### LINKED ISSUE #235: add startup commands to data service
default - always cache

--cache-duration int (seconds) default: 300

- data service should be able to run in multiple ways for testing
  - --in-memory bool
    - currently exists
  - --no-cache bool (never cache)
    - currently exists
  - --perf-cache int 
    - only use cache when req / sec > --perf-cache

Validation:
- --cache-duration - int > 0
- --perf-cache - int > 0
- only one of the three options can be specified

### PR DESCRIPTION
**Pull request by bartr**
# Design Review Scheduled

## Type of PR

- [ ] Documentation changes
- [x] Code changes
- [x] Test changes
- [ ] CI-CD changes
- [ ] GitHub Template changes

## Purpose of PR
- added new command line options to control cache

## Validation

- [x] Unit tests updated and ran successfully
- [x] Update documentation or issue referenced above

## Issues Closed or Referenced

- Closes #235

### CODE REVIEW

File: spikes/data/dataservice/app/Core/CommandLine.cs [hunk 1]
```diff
@@ -29,19 +34,16 @@ public static string[] CombineEnvVarsWithCommandLine(string[] args)
 
             List<string> cmd = new List<string>(args);
 
-            // add --log-level value from environment or default
-            if (!cmd.Contains("--log-level") && !cmd.Contains("-l"))
-            {
-                string logLevel = Environment.GetEnvironmentVariable("LOG_LEVEL");
+            // add values from environment
```
Comment (bartr) [inline spikes/data/dataservice/app/Core/CommandLine.cs:37]: new extension to simplify code

File: spikes/data/dataservice/app/Core/CommandLine.cs [hunk 2]
```diff
@@ -54,17 +56,23 @@ public static RootCommand BuildRootCommand()
...
             {
-                Name = "ngsa",
-                Description = "Next Gen Symmetric App",
+                Name = "DataService",
+                Description = "NGSA Data Service",
                 TreatUnmatchedTokensAsErrors = true,
             };
 
             // add the options
+            root.AddOption(new Option<int>(new string[] { "--cache-duration" }, () => 300, "Cache for duration (seconds)"));
```
Comment (bartr) [inline spikes/data/dataservice/app/Core/CommandLine.cs:65]: new options

File: spikes/data/dataservice/app/Core/CommandLine.cs [hunk 3]
```diff
@@ -75,54 +83,22 @@ public static RootCommand BuildRootCommand()
...
         /// <param name="inMemory">Use in-memory DB</param>
+        /// <param name="noCache">don't cache results</param>
+        /// <param name="perfCache">cache results under load</param>
+        /// <param name="cacheDuration">cache duration (seconds)</param>
         /// <returns>status</returns>
-        public static async Task<int> RunApp(string secretsVolume, LogLevel logLevel, bool dryRun, bool inMemory)
+        public static async Task<int> RunApp(string secretsVolume, LogLevel logLevel, bool dryRun, bool inMemory, bool noCache, int perfCache, int cacheDuration)
         {
             try
             {
```
Comment (bartr) [inline spikes/data/dataservice/app/Core/CommandLine.cs:93]: this code is moved to a method below

File: spikes/data/dataservice/app/Core/CommandLine.cs [hunk 4]
```diff
@@ -75,54 +83,22 @@ public static RootCommand BuildRootCommand()
...
-                    if (ndx > 0)
-                    {
-                        CosmosName = CosmosName.Remove(ndx);
-                    }
-                }
-
-                // setup ctl c handler
-                ctCancel = SetupCtlCHandler();
-
+                // assign command line values
```
Comment (bartr) [inline spikes/data/dataservice/app/Core/CommandLine.cs:94]: track values as static at App.*

File: spikes/data/dataservice/app/Core/CommandLine.cs [hunk 5]
```diff
@@ -145,20 +121,23 @@ public static async Task<int> RunApp(string secretsVolume, LogLevel logLevel, bo
                     return -1;
                 }
 
-                // don't start the web server
+                // display dry run message
                 if (dryRun)
                 {
                     return DoDryRun();
                 }
```
Comment (bartr) [inline spikes/data/dataservice/app/Core/CommandLine.cs:129]: moved from above

File: spikes/data/dataservice/app/Core/CommandLine.cs [hunk 6]
```diff
@@ -182,22 +161,139 @@ public static async Task<int> RunApp(string secretsVolume, LogLevel logLevel, bo
             }
         }
 
-        /// <summary>
-        /// Display the dry run message
-        /// </summary>
-        /// <param name="authType">authentication type</param>
-        /// <returns>0</returns>
```
Comment (bartr) [inline spikes/data/dataservice/app/Core/CommandLine.cs:189]: code from above

File: spikes/data/dataservice/app/Core/CommandLine.cs [hunk 7]
```diff
@@ -182,22 +161,139 @@ public static async Task<int> RunApp(string secretsVolume, LogLevel logLevel, bo
...
+
+                int ndx = CosmosName.IndexOf('.', StringComparison.OrdinalIgnoreCase);
+
+                if (ndx > 0)
+                {
+                    CosmosName = CosmosName.Remove(ndx);
+                }
+            }
+        }
+
```
Comment (bartr) [inline spikes/data/dataservice/app/Core/CommandLine.cs:207]: new command line validation code

File: spikes/data/dataservice/app/Core/CommandLine.cs [hunk 8]
```diff
@@ -182,22 +161,139 @@ public static async Task<int> RunApp(string secretsVolume, LogLevel logLevel, bo
...
+            // return error message(s) or string.empty
+            return msg;
+        }
+
+        // Display the dry run message
         private static int DoDryRun()
         {
             Console.WriteLine($"Version            {Middleware.VersionExtension.Version}");
             Console.WriteLine($"Log Level          {AppLogLevel}");
```
Comment (bartr) [inline spikes/data/dataservice/app/Core/CommandLine.cs:286]: added new command line values to --dry-run

File: spikes/data/dataservice/app/Core/CommandLineExtensions.cs [hunk 9]
```diff
@@ -0,0 +1,32 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (bartr) [inline spikes/data/dataservice/app/Core/CommandLineExtensions.cs:1]: new extension to handle merging env vars with command line

File: spikes/data/dataservice/app/DataAccessLayer/InMemoryDal.cs [hunk 10]
```diff
@@ -19,10 +19,6 @@
 /// </summary>
 namespace CSE.NextGenSymmetricApp.DataAccessLayer
 {
```
Comment (bartr) [inline spikes/data/dataservice/app/DataAccessLayer/InMemoryDal.cs:21]: moved to stylecop.json

File: spikes/data/dataservice/app/DataAccessLayer/InMemoryDal.cs [hunk 11]
```diff
@@ -85,12 +81,12 @@ public InMemoryDal()
             }
 
             // load the data from the json file
```
Comment (bartr) [inline spikes/data/dataservice/app/DataAccessLayer/InMemoryDal.cs:83]: converted to system.text.json - dynamic doesn't work so created a model class

issue on the board to fix naming - it has dependencies

File: spikes/data/dataservice/app/DataAccessLayer/InMemoryDal.cs [hunk 12]
```diff
@@ -146,6 +142,40 @@ public Task<IEnumerable<Actor>> GetActorsAsync(ActorQueryParameters actorQueryPa
             });
         }
```
Comment (bartr) [inline spikes/data/dataservice/app/DataAccessLayer/InMemoryDal.cs:144]: added get actor IDs to "search service"

File: spikes/data/dataservice/app/DataAccessLayer/Model/GenreObject.cs [hunk 13]
```diff
@@ -0,0 +1,13 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
```
Comment (bartr) [inline spikes/data/dataservice/app/DataAccessLayer/Model/GenreObject.cs:3]: new genre model

File: spikes/data/dataservice/app/DataAccessLayer/dalActors.cs [hunk 14]
```diff
@@ -15,11 +15,6 @@ namespace CSE.NextGenSymmetricApp.DataAccessLayer
     /// </summary>
     public partial class CosmosDal
     {
```
Comment (bartr) [inline spikes/data/dataservice/app/DataAccessLayer/dalActors.cs:17]: query info replaced by "search service"

File: spikes/data/dataservice/app/DataAccessLayer/dalMovies.cs [hunk 15]
```diff
@@ -66,17 +66,15 @@ public async Task<IEnumerable<Movie>> GetMoviesAsync(MovieQueryParameters movieQ
 
             string ids = App.SearchService.GetMovieIds(movieQueryParameters);
```
Comment (bartr) [inline spikes/data/dataservice/app/DataAccessLayer/dalMovies.cs:68]: removed completed TODOs

cache empty result per design review

File: spikes/data/dataservice/app/Program.cs [hunk 16]
```diff
@@ -52,6 +53,10 @@ public sealed partial class App
         /// Gets or sets LogLevel
         /// </summary>
         public static LogLevel AppLogLevel { get; set; } = LogLevel.Warning;
```
Comment (bartr) [inline spikes/data/dataservice/app/Program.cs:55]: static values from command line

File: spikes/data/dataservice/tests/AppTests.cs [hunk 17]
```diff
@@ -18,7 +18,7 @@ public async Task RunApp()
             Assert.Equal(0, await App.Main(args));
 
             args = new string[] { "--secrets-volume", "foo" };
```
Comment (bartr) [inline spikes/data/dataservice/tests/AppTests.cs:20]: error is caught in system.commandline which returns 1

File: spikes/data/dataservice/tests/AppTests.cs [hunk 18]
```diff
@@ -38,7 +38,8 @@ public void CommandLineTests()
             Assert.Equal(0, root.Parse("-d").Errors.Count);
             Assert.Equal(0, root.Parse("-l Error").Errors.Count);
             Assert.Equal(1, root.Parse("-l foo").Errors.Count);
-            Assert.Equal(0, root.Parse("--secrets-volume foo").Errors.Count);
```
Comment (bartr) [inline spikes/data/dataservice/tests/AppTests.cs:41]: system.commandline validates the secrets volume exists

**Review by jofultz — APPROVED** [review]

**Pull request merged**